unescape: fix use of uninitialized variable if string starts with \'

l2dy opened the pull request:
In `flb_unescape_string_utf8`, if the in_buf string starts with an esacped
single quote (\\'), the ch variable would be uninitialized when used in
`esc_out = u8_wc_toutf8(temp, ch);`.

l2dy commented:
Alternatively, we could remove `'\''` from is_json_escape().

nigels-com commented:
The change looks legit to me.

Seems to me that the `\a` and `\v` cases are unreachable code according to the logic of `is_json_escape`.

nigels-com commented:
According to [RFC 4627](https://www.ietf.org/rfc/rfc4627.txt):

> A string begins and ends with
   quotation marks.  All Unicode characters may be placed within the
   quotation marks except for the characters that must be escaped:
   quotation mark, reverse solidus, and the control characters (U+0000
   through U+001F).

The implementation here appears to be intended to handle some of the control characters `00` through `1f` - in `\a` (etc) form. [wikipedia](https://en.wikipedia.org/wiki/Control_character#In_ASCII)

bluebike commented:
Other problems in original code...

In function **flb_unescape_string_utf8** 

https://github.com/fluent/fluent-bit/blob/3e5645283603622168a3f29dcf2bbd8312bc8164/src/flb_unescape.c#L155

Calling **is_json_escape** could be removed... if all cases were handled in switch.

In **u8_read_escape_sequence** 
https://github.com/fluent/fluent-bit/blob/3e5645283603622168a3f29dcf2bbd8312bc8164/src/flb_unescape.c#L70

Problem is that  this can read past of string buffer when we 
have bad utf8-string at the end of source string. 
Bound check is missing.

edsiper commented:
folks, to get this merged, would you please update the PR with the latest fixes/suggestions shared by the community ?

l2dy commented:
> folks, to get this merged, would you please update the PR with the latest fixes/suggestions shared by the community ?

OK, removed the redundant `is_json_escape` call (also removed the `\a` and `\v` cases for API compatibility with original implementation).

I think the `u8_read_escape_sequence` issue should be fixed in a different pull request though.

edsiper commented:
thanks

edsiper commented:
can you help us with the remaining issue ?

l2dy commented:
> In **u8_read_escape_sequence**
> 
> https://github.com/fluent/fluent-bit/blob/3e5645283603622168a3f29dcf2bbd8312bc8164/src/flb_unescape.c#L70
> 
> Problem is that this can read past of string buffer when we
> have bad utf8-string at the end of source string.
> Bound check is missing.

If the string has a malformed escape sequence at the end, *octal_digit* or *hex_digit* would break the loop and won't read ahead.

If the string ends with a `\`, `*next` should be `'\0'`, then all the *if* matches in *u8_read_escape_sequence* would fail and return 1, with `ch = 0`.

I think it can't read past the string buffer. Did I miss something?

yann-soubeyrand commented:
@l2dy from my understanding, in_buf is not necessarily a null byte terminated string. It seems to be, at least in some cases, a buffer containing sz characters (without a null byte at the end). Therefore, switch(next) could access memory past the buffer end, which can be easily avoided with a bound check. But it would be more problematic for the call to u8_read_escape_sequence which doesn't do bound checks as mentioned earlier by @bluebike. We should just write tests and execute them using Vlagrind ;-)

l2dy commented:
Thanks for the explanation. In this case, the `*next` expression here could also read past the buffer by one byte.

https://github.com/fluent/fluent-bit/blob/262267287641d6d54bb62f0a0ada1c76a5bd40b4/src/flb_unescape.c#L136-L141

l2dy commented:
> can you help us with the remaining issue ?

Sorry, but I don't have time to work on this right now. I've created a new issue to track the buffer over-read. #1582
